### PR TITLE
class.c: Define a :writer attribute, applicable to scalar fields only

### PR DESCRIPTION
**Pull request by leonerd**
Permits automatic creation of writer accessor methods, in addition to the reader ones we already had.

---------------------------------------------------------------------------------
* This set of changes requires a perldelta entry, and it is included.

### CODE REVIEW

**Review by rwp0 — APPROVED** [review]

File: class.c [hunk 1]
```diff
@@ -1052,6 +1052,101 @@ apply_field_attribute_reader(pTHX_ PADNAME *pn, SV *value)
...
+    op->op_private = idx;
+    return op;
+}
+
+static void
+apply_field_attribute_writer(pTHX_ PADNAME *pn, SV *value)
+{
+    char sigil = PadnamePV(pn)[0];
+    if(sigil != '$')
+        croak("Cannot apply a :writer attribute to a non-scalar field");
```
Comment (tonycoz) [inline class.c:1075]: No test for this error.

**Comment by tonycoz** [comment]
It makes deparse unhappy:
```
$ ./perl -Ilib -MO=Deparse,Foo::set_x -Mfeature=class -e 'class Foo { field $x :writer; }'
class is experimental at -e line 1.
field is experimental at -e line 1.
use feature 'class';
{
    package Foo;
    ;
}
unexpected OP_METHSTART at lib/B/Deparse.pm line 1677.
package Foo;
sub set_x {
    XXX;
    die sprintf("Too many arguments for subroutine at %s line %d.\n", (caller)[1, 2]) unless @_ <= 1;
    die sprintf("Too few arguments for subroutine at %s line %d.\n", (caller)[1, 2]) unless @_ >= 1;
    $x = $_[0];
    return $self;
}
-e syntax OK
```
but that looks like a general deparse problem with methods:
```
$ ./perl -Ilib -MO=Deparse,Foo::set_x -Mfeature=class -e 'class Foo { field $x; method set_x { $x = $_[0]; } }'
class is experimental at -e line 1.
field is experimental at -e line 1.
method is experimental at -e line 1.
unexpected OP_METHSTART at lib/B/Deparse.pm line 1677.
use feature 'class';
{
    package Foo;
    ;
    sub set_x : method {
        XXX;
        $x = $_[0];
    }
}
-e syntax OK
```

You asked about identifier validation in #p5p, but:
```
$ ./perl -Ilib -Mfeature=class -E 'class Foo { field $x :reader :writer(a c); } my $m = "a c"; say Foo->new()->$m("abc")->x'
class is experimental at -e line 1.
field is experimental at -e line 1.
abc
```

but I expect :reader has the same problem.

Similarly:
```
$ ./perl -Ilib -Mfeature=class -E 'class Foo { field $x :writer( ); } say "<$_>" for keys %Foo::'
class is experimental at -e line 1.
field is experimental at -e line 1.
<__ANON__>
<new>
```
(no method defined, but no error reported)

**Comment by leonerd** [comment]
On grounds that a bunch of these issues also affect the `:reader` attribute as well, I think I will first go fix them in a separate PR before this, and then come back to this one after that's done.

**Comment by leonerd** [comment]
Now with a guard against invalid method names, and a test for the non-scalar error message.

The deparse issues remain, but that also applies to existing methods in general, including `:reader` generated ones. We should look at that later.

**Review by tonycoz — APPROVED** [review]

File: pod/perlclass.pod [hunk 2]
```diff
@@ -260,6 +260,33 @@ context.
...
+By default the accessor method will have the name of the field minus the
+leading sigil with the string C<set_> prefixed to it, but a different name
+can be specified in the attribute's value.
+
+    field $x :writer(write_x);
+
+    # Generates a method
+    method write_x ($new) { ... }
+
+Curerently, writer accessors can only be applied to scalar fields.  Attempts
```
Comment (esabol) [inline pod/perlclass.pod:285]: "Currently" is misspelled on this line.

File: pod/perlclass.pod [hunk 3]
```diff
@@ -260,6 +260,33 @@ context.
 
     scalar $instance->users;
 
+=head3 :writer
+
+A field with a C<:writer> attribute will generate a writer accessor method
+automatically.  The generated method will have a signature that consumes
+exactly one argument, and its body will assign that scalar argument to the
+field, and return the invocant object itself.
```
Comment (esabol) [inline pod/perlclass.pod:268]: The comma on this line (268) is incorrect and should be removed.

**Review by esabol — COMMENTED** [review]
I know this has already been merged, but I hope someone will be able to correct these anyway. :smile:

File: pod/perlclass.pod (hunk 3 above)
Comment (jkeenan) [reply pod/perlclass.pod:268]: Thanks.  Committed to blead in 9e152dfc44.

**Pull request merged**